### PR TITLE
Components with modificators - internal, public, private using Razor syntax 2

### PR DESCRIPTION
**Pull request by SamProf**
Summary of the changes
- Implementing possibility to use directives like `@public`, `@internal`, `@private`
- Removed checking of accessibility of component type in checking type is Component
- Added example how use `@internal` component

Issue: Provide a way to create private Components using Razor syntax dotnet/razor-tooling#7051 

This is first inital solution, just want to know, @SteveSandersonMS is this a right way to solve this problem.

### CODE REVIEW

**Comment by ajaybhargavb** [comment]
> @ajaybhargavb does this affect discovery in tooling?

Right now, non-public types won't even be recognized as components. See https://github.com/dotnet/aspnetcore/blob/master/src/Razor/Microsoft.CodeAnalysis.Razor/src/ComponentDetectionConventions.cs#L23. 

I don't think tooling checks accessibility modifiers anywhere. If it is present in the serialized json, it should pick it up. Correct me if I am wrong, @NTaylorMullen

**Comment by pranavkm** [comment]
Is adding support for non-public types in tooling difficult? I vaguely recall one of the reasons we removed support for non-public properties was because you had to do write complicated code for the non-public symbols to be resolved.

**Comment by ajaybhargavb** [comment]
> Is adding support for non-public types in tooling difficult?

I wasn't sure so I was digging through PRs from history but found that [you added](https://github.com/aspnet/Razor/pull/2714/files#diff-0c1bdeabe1b9fd0514236bb87b513bb3R126) this part of the code 😆. So maybe your recall is a better guess than mine.

**Comment by NTaylorMullen** [comment]
> I don't think tooling checks accessibility modifiers anywhere. If it is present in the serialized json, it should pick it up. Correct me if I am wrong, @NTaylorMullen

You're  100 %  correct. With how t his is currently  the tooling will treat  all components as public. It's something that would have to be  added  as   part of  TagHelper  discovery  and then consumed by tooling to properly allow the correct TagHelpers for various  projects.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/Extensions/AccessibilityModificatorDirectives.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            "public",
+            DirectiveKind.SingleLine,
+            builder =>
+            {
+                builder.Usage = DirectiveUsage.FileScopedMultipleOccurring;
+                builder.Description = "Add public access modifier for component type";
+            });
+
+        public static readonly DirectiveDescriptor PrivateDirective = DirectiveDescriptor.CreateDirective(
+            "private",
```
Comment (NTaylorMullen) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/Extensions/AccessibilityModificatorDirectives.cs:26]: What would private mean for Razor based components?

**Comment by mkArtakMSFT** [comment]
Thanks for your contribution. We are closing this PR as we plan to suggest a design related to how this can be handled in the related issue. We will welcome contributions after that.

**Pull request closed without merging**